trie/smt.py

```diff
@@ -0,0 +1,266 @@
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
+            if len(node_updates) <= branch_point:
+                raise ValidationError("Updated node list is not deep enough")
+
+            # Update sibling in the branch where our key differs from the update
+            self.branch[branch_point] = node_updates[branch_point]
+            # No need to update leaf
+
+
+class SparseMerkleTree:
+
+    def __init__(self, keysize=32, default=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= keysize <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self.keysize = keysize  # keysize in bytes
+        self.depth = keysize * 8  # depth is number of bits in key
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
+        self.db = {}
+        node = self._default  # Default leaf node
+        for i in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
+            node = node_hash + node_hash
+
+        # Finally, write the root hash
+        self.root_hash = keccak(node)
+        self.db[self.root_hash] = node
+
+    @classmethod
+    def from_db(cls, db, root_hash, keysize=32, default=BLANK_NODE):
+        smt = cls(keysize=keysize, default=default)
+
+        # If db is provided, and is not consistent,
+        # there may be a silent error. Can't solve that easily.
+        for k, v in db.items():
+            validate_is_bytes(k)
+            validate_length(k, 32)  # Must be a bytes32 hash
+            validate_is_bytes(v)
+            smt.db[k] = v
+
+        # Set root_hash, so we know where to start
+        validate_is_bytes(root_hash)
+        validate_length(root_hash, 32)  # Must be a bytes32 hash
+
+        smt.root_hash = root_hash
+
+        return smt
+
+    def get(self, key):
+        value, _ = self._get(key)
+
+        # Ensure that it isn't blank!
+        if value == BLANK_NODE:
+            raise KeyError("Key does not exist")
+
+        return value
+
+    def branch(self, key):
+        value, branch = self._get(key)
+
+        # Ensure that it isn't blank!
+        if value == BLANK_NODE:
+            raise KeyError("Key does not exist")
+
+        return branch
+
+    def _get(self, key):
+        """
+        Returns db value and branch in root->leaf order
+        """
+        validate_is_bytes(key)
+        validate_length(key, self.keysize)
+        branch = []
+
+        target_bit = 1 << (self.depth - 1)
+        path = to_int(key)
+        node_hash = self.root_hash
+        # Append the sibling to the branch
+        # Iterate on the parent
+        for i in range(self.depth):
```

Comment:
We have tended to use `_` for loop variables that we don't actually make use of.

Reply:
Will do